dimod/serialization/lp.py

```diff
@@ -0,0 +1,268 @@
+    left_ineq = number_or_inf + sense
+    right_ineq = sense + number_or_inf
+    bounded_var = Group(Optional(left_ineq).setResultsName("leftbound") + valid_name +
+                        Optional(right_ineq).setResultsName("rightbound"))
+    free_var = Group(valid_name + Literal("free"))
+
+    bounds_word = free_var | bounded_var
+    bounds = bounds_tag + ZeroOrMore(bounds_word).setResultsName("bounds")
+
+    # generals (integer variables)
+    generals = gen_tag + ZeroOrMore(valid_name).setResultsName("generals")
+
+    # binaries (binary variables)
+    binaries = bin_tag + ZeroOrMore(valid_name).setResultsName("binaries")
+
+    var_info = ZeroOrMore(bounds | generals | binaries)
+
+    # full LP file grammar
+    grammar = objective + constraints_tag + constraints + var_info + end_tag
+
+    # commenting
+    comment_style = Literal("\\") + restOfLine
+    grammar.ignore(comment_style)
+
+    return grammar
+
+
+def remove_strings(string, strings_to_remove):
+    """Replaces an iterable of strings in removables
+        if removables is a string, each character is removed """
+    for r in strings_to_remove:
+        try:
+            string = string.replace(r, "")
+        except TypeError:
+            raise TypeError("Strings_to_remove contains a non-string element")
+    return string
+
+
+def get_variables(parse_output: ParseResults,
+                  lower_bound_default: typing.Optional[int] = None,
+                  upper_bound_default: typing.Optional[int] = None) -> Dict:
+    """
+    Returns set of variables and dict of variable types and bounds
+    Args:
+        parse_output: the parse results encoding the LP file
+        lower_bound_default: the lower bound of the integer variables, in case they are not specified
+        upper_bound_default: the upper bound of the integer variables in case they are not specified
+
+    Returns:
+        a dictionary for each variable containing type, upper and lower bounds
+    """
+    # handle the default bounds
+    if lower_bound_default is None:
+        lb_def = -infinity
+    else:
+        lb_def = lower_bound_default
+
+    if upper_bound_default is None:
+        ub_def = infinity
+    else:
+        ub_def = upper_bound_default
+
+    all_vars = set()
+    # default variable type for LP file
+    # should be continuous, even though
+    # they are not supported in CQMs
+    variables_info = defaultdict(lambda: {"type": "c", "lb": lb_def, "ub": ub_def})
+
+    # scan the objective
+    for oe in parse_output.objective:
+        if isinstance(oe, str):
+            continue
+
+        else:
+            if len(oe) == 2:
+                if oe.name != "":
+                    all_vars.add(oe.name[0])
+                else:
+                    # not supported yet.. although they can be replaced by CQM function
+                    raise ValueError("pure quadratic terms are not supported, found {}^2".format(oe.squared_name[0]))
```

Comment:
I believe this is saying we don't support `x^2` or `i^2` etc. But we actually do in the QM. Specifically, you can use the `x*x == x` trick for binary, and we do support integer squares.